File: src/scrapers/strategies/computrabajo.ts
import { Job, ScraperConfig, ScraperResult } from '../types';
import { PuppeteerScraper } from './puppeteerScraper';

/**
 * Computrabajo.com job scraper
 * Uses Puppeteer with stealth for headless Chrome automation
 * Popular in Latin America and Spain
 */
export class ComputrabajoScraper extends PuppeteerScraper {
  private readonly baseUrl = 'https://www.computrabajo.com';

  constructor(config: ScraperConfig) {
    super(config);
  }

  /**
   * Build Computrabajo job search URL
   * @param config - Scraper configuration
   * @returns Encoded search URL
   */
  protected buildUrl(config: ScraperConfig): string {
    const query = encodeURIComponent(config.query || '');
    // Search for jobs from last 3 days
    return `${this.baseUrl}/td/ofertas?q=${query}&asignar=1&f_date=3`;
  }

  /**
   * Wait for job listings to appear
   * Computrabajo-specific selectors
   * @param page - Puppeteer page
   */
  protected async waitForJobs(page: any): Promise<void> {
    try {
      // Wait for job listing containers
      await Promise.race([
        page.waitForSelector('.offer-box', { timeout: 10000 }),
        page.waitForSelector('.oferta', { timeout: 10000 }),
        page.waitForSelector('[data-tn-element="jobCard"]', { timeout: 10000 }),
      ]);
      (this as any).logger.info('Job listings found on Computrabajo');
    } catch (error) {
      (this as any).logger.warn('Job selector timeout - Computrabajo may have anti-bot measures');
    }
  }

  /**
   * Extract job data from the page
   * Computrabajo-specific extraction logic
   * @param page - Puppeteer page
   * @returns Array of extracted jobs
   */
  protected async extractJobs(page: any): Promise<Job[]> {
    return await page.evaluate((): Job[] => {
      const jobs: Job[] = [];

      // Try multiple selector strategies
      const jobSelectors = [
        '.offer-box',
        '.oferta',
        '[data-tn-element="jobCard"]',
        '.job-card',
      ];

      let jobElements: Element[] = [];
      for (const selector of jobSelectors) {
        const elements = document.querySelectorAll(selector);
        if (elements.length > 0) {
          jobElements = Array.from(elements);
          break;
        }
      }

      jobElements.forEach((el: Element) => {
        try {
          const titleEl = el.querySelector('.offer-title a, .oferta-titulo a, h2 a');
          const companyEl = el.querySelector('.offer-company, .empresa, .company-name');
          const locationEl = el.querySelector('.offer-branch, .lugar, .location');
          const linkEl = el.querySelector('a[href*="/trabajo-"]') || el.querySelector('a');
          const dateEl = el.querySelector('.offer-date, .fecha, [data-date]');

          const title = titleEl?.textContent?.trim() || '';
          const company = companyEl?.textContent?.trim() || '';
          const location = locationEl?.textContent?.trim() || '';
          const link = linkEl?.getAttribute('href') || '';
          const dateText = dateEl?.textContent?.trim() || '';

          // Skip if essential fields are missing
          if (!title || !company) return;

          // Normalize URL
          const fullLink = link?.startsWith('http') 
            ? link 
            : `https://www.computrabajo.com${link?.startsWith('/') ? link : '/' + link}`;

          jobs.push({
            id: '',
            title,
            company,
            location,
            link: fullLink,
            description: '',
            source: 'computrabajo',
            scrapedAt: new Date(),
          });
        } catch (error) {
          console.warn('Error parsing Computrabajo job card:', error);
        }
      });

      return jobs;
    });
  }
}

File: scrapers/computrabajo.py
from scrapers.shared.base import ScraplingBaseScraper, run_scraper_cli
from scrapers.shared.models import Job


class ComputrabajoScraper(ScraplingBaseScraper):
    name = "computrabajo"

    def build_url(self, query: str) -> str:
        base = (self.config.extra or {}).get("base_url", "https://co.computrabajo.com")
        slug = query.lower().replace(" ", "-")
        return f"{base}/trabajo-de-{slug}?q={query}"

    def _wait_selector(self) -> str | None:
        return None

    def _fetcher_options(self) -> dict:
        return {"network_idle": False, "wait": 3000}

    def parse_page(self, page) -> list[Job]:
        jobs: list[Job] = []
        job_elements = page.css("article.box_offer")

        if not job_elements:
            job_elements = page.css("[class*='box_offer']")

        self._log("info", f"Found {len(job_elements)} offer elements")

        for el in job_elements:
            job = self._parse_card(el)
            if job:
                jobs.append(job)

        return jobs

    def _parse_card(self, el) -> Job | None:
        try:
            title = ""
            title_els = el.css("a.js-o-link")
            if title_els:
                title = title_els[0].text.strip() if title_els[0].text else ""
            if not title:
                title_els = el.css("h2 a, h3 a")
                if title_els:
                    title = title_els[0].text.strip() if title_els[0].text else ""

            company = ""
            company_els = el.css("a.fc_base.t_ellipsis")
            if company_els:
                company = company_els[0].text.strip() if company_els[0].text else ""
            if not company:
                company_els = el.css("p.dFlex.vm_fx.fs16.fc_base.mt5")
                if company_els:
                    company = company_els[0].text.strip() if company_els[0].text else ""

            location = ""
            loc_els = el.css("span.mr10")
            if loc_els:
                location = loc_els[0].text.strip() if loc_els[0].text else ""

            link = ""
            link_els = el.css("a.js-o-link")
            if link_els:
                link = link_els[0].attrib.get("href", "")

            if not title:
                return None

            if company.lower() == "computrabajo":
                return None

            if link and not link.startswith("http"):
                base = (self.config.extra or {}).get("base_url", "https://co.computrabajo.com")
                prefix = "" if link.startswith("/") else "/"
                link = f"{base}{prefix}{link}"

            return Job(
                id=self._generate_id(link),
                title=title,
                company=company,
                location=location,
                link=link,
                description="",
                source="computrabajo",
            )
        except Exception as exc:
            self._log("warning", f"Error parsing card: {exc}")
            return None

        return jobs


if __name__ == "__main__":
    run_scraper_cli(ComputrabajoScraper)
